### ISSUE TITLE
embedding ffmpeg and imagemagick?

### ISSUE THREAD

**Issue by gre** [nwjs/nw.js#3355]
Have anyone tried to use ffmpeg and imagemagick in a nw.js app, and especially I would like to embed them in the app (so a windows/mac user don't have to install them from CLI).
Is that even possible?

Thanks

**Comment by MarcFowler** [comment]
Yes, I've done this. You need to build ffmpeg statically, and then you can use something like https://github.com/fluent-ffmpeg/node-fluent-ffmpeg for instance to interact with it. You just provide the path to it, which can be inside your app (e.g. `./bin/osx/ffmpeg` as the path to it).

The hardest part is getting that static build. There are pre-built binaries out there (http://ffmpegmac.net/ for instance) but they might not have all the functionality you need.

fwiw you can use any CLI binary with nw.js by using node's https://nodejs.org/api/child_process.html module to spawn and control it/read it's output/etc.

**Comment by MarcFowler** [comment]
Oh and re. imagemagick specifically, I used graphicsmagick because it was simpler to compile statically!

**Comment by gre** [comment]
Is there any guide to build and embed native libraries (like graphicsmagick) into a nw.js application for Linux / Mac / Windows or any tool you would recommend for that?

I have found https://github.com/nwjs/nw.js/wiki/How-to-package-and-distribute-your-apps but it is not very detailed.

Thanks

**Comment by MarcFowler** [comment]
I don't know of a guide but it is essentially very simple..

1. Compile a static build of the library you want (such as GM). This is by far the most complicated part because you may need to statically compile dependencies like `libpng` for instance first, and then provide those to the linker for GM.

2. Put your binary into your app's folder (`bin/osx/ffmpeg` for instance)

3. Start using the binary in your code; for example:

```
var child_process = require('child_process');
child_process.execFile('./bin/osx/ffmpeg', [], function() {
  // deal with the response here
});
```

For this part you'll want to be reading up on [the node documentation for child processes](https://nodejs.org/api/child_process.html#child_process_child_process) and knowing whether you want to execute a command and get a single response, or to spawn a long-life process which will run in the background and return data to you as it gets things. The first would be an example of running a command to delete a file, but the second would be to convert an image or video, because you'd want to get progress as it goes through the conversion and display that most likely.

You can find some pre-built binaries, such as [for ffmpeg on OSX](https://evermeet.cx/ffmpeg/). They are usually easier to find for Windows though, such as [Zeranoe's builds.](http://ffmpeg.zeranoe.com/builds/) But by and large you'll want to compile your own. It is easy and 99% of projects explain how to do it in their documentation.

I hope that helps!

**Comment by armoucar** [comment]
Glad I finally found this. :)

Thanks @MarcFowler

**Comment by rogerwang** [comment]
Thanks. And I believe it's a valid feature request.

**Comment by Connum** [comment]
+1 for using ffmpeg directly via an API as its already built in (without proprietary codecs).
